File: trains.js
#!/usr/bin/env node
/*
 trains.js
(c) 2019 Mal Minhas, <mal@malm.co.uk>

Licence
--------
Copyright 2019 Mal Minhas. All Rights Reserved.

Description
-----------
transportAPI is documented here: https://developer.transportapi.com/docs?raml=https://transportapi.com/v3/raml/transportapi.raml
More on three letter train codes here: http://www.railwaycodes.org.uk/crs/CRS0.shtm
Some examples are:
1. London Paddington = PAD
2. London Waterloo = WAT
3. London Bridge = LD

Installation
------------
$ npm install node-fetch
$ npm install docopt

Version
-------
11.06.19  0.1   Initial version based on https://github.com/chrishutchinson/train-departure-screen/blob/master/src/trains.py
12.06.19  0.2   Added docopt support
*/ 

'use strict';

const fs = require('fs')
const fetch = require('node-fetch');
// Needed for older node versions.  
// See: https://stackoverflow.com/questions/52566578/url-is-not-defined-in-node-js
const URL = require('url').URL;

const PROGRAM = 'trains.js'
const VERSION = '0.2'
const DATE = '11.06.19'
const AUTHOR = 'Mal Minhas'

const APP_ID = readCred('.transportAppId')
const APP_KEY = readCred('.transportAppKey')

function readCred(fname) {
    return fs.readFileSync(fname, 'utf8');
}

function getTrainsCallingAt(station_code,dest_code,dest_name) {
    let url = new URL(`http://transportapi.com/v3/uk/train/station/${station_code}/live.json`)
    let params = {  'app_id': APP_ID, 
                    'app_key': APP_KEY, 
                    //'station_name':station_name, 
                    'station_code':station_code, 
                    'calling_at': dest_code, 
                    'type': 'departure'}
    Object.keys(params).forEach(key => url.searchParams.append(key, params[key]))
    const urlstring = url.toString()
    //console.log(urlstring)
    fetch(urlstring)
      .then(response => response.json())
      .then(data => {
        data['destination_code'] = dest_code
        data['destination_name'] = dest_name    
        formatTrains(data)
      })
      .catch(() => {
        console.log('Caught issue with getTrainsCallingAt fetch url')
      })
}

function procStop(stop) {
    let s = {}
    s.station_code = stop.station_code
    s.station_name = stop.station_name
    s.expected_arrival_time = stop.expected_arrival_time
    s.platform = stop.platform
    return s
}

function getStops(train,d) {
    // We need to make a GET request on the timetable URL to retrieve array of stops.  
    // We are then only interested in a subset of the details fields for each stop.
    // We also want to add whether that stop is on the designated journey or not.

    let arr = []
    let timetable_url = train.service_timetable.id
    fetch(timetable_url)
        .then(response => response.json())
        .then(data => {
            let stops = data.stops
            let on_route = false
            stops.forEach(stop => {
                const s = procStop(stop)
                if (s.station_code === d.station_code) {
                    s.on_route = true
                    on_route = true
                } else if (s.station_code === d.dest_code) {
                    s.on_route = true
                    on_route = false
                } else {
                    s.on_route = on_route
                }
                arr.push(s)
            })
            let trainDetails = formatDeparture(train,arr,d)
            printTrainDetails(trainDetails,arr)
        })
        .catch(() => {
            console.log('Caught issue with getStops fetch url')
        })    
}

function formatHeader(d){
    let header = `==== Trains from ${d.station_name} (${d.station_code}) to ${d.destination_name} `
    header += `(${d.destination_code}) ${d.time_of_day} ${d.date} ====`
    return header
}

function formatDeparture(train,stops,d){
    let source = stops.filter(stop => stop.station_code === d.station_code)[0]
    let dest = stops.filter(stop => stop.station_code === d.destination_code)[0]
    let route = stops.filter(stop => stop.on_route === true)

    let departure = `${d.station_code} ${train.expected_departure_time} -> ${d.destination_code}`
    departure += ` ${dest.expected_arrival_time} => ${train.status}\n`
    departure += `\tTrain ${train.train_uid} (${train.operator}) from ${train.origin_name}`
    departure += ` arriving at ${d.station_name} on platform ${source.platform}`
    departure += ` going to ${d.destination_name} platform ${dest.platform}. ${route.length} stops:`
    return departure
}

function formatTrains(data) {
    // Keys: 'date', 'time_of_day', 'request_time', 'station_name', 'station_code', 'departures'
    // where 'departures' is a dict with one key 'all' which is a list of dicts of train departures
    const header = formatHeader(data)
    printHeader(header)
    let trains = data.departures.all
    trains.forEach(train => {
        getStops(train, data)
    })
}

function printHeader(header){
    console.log('='.repeat(header.length))
    console.log(header)
    console.log('='.repeat(header.length))
}

function printStopNames(stops){
    let stopNames = []
    let stopsOnRoute = stops.filter(stop => stop.on_route === true)
    stopsOnRoute.forEach(stop => {
        stopNames.push(stop.station_name)
    })
    console.log(`\t${stopNames.join(',')}`)
}

function printTrainDetails(trainDetails,stops){
    console.log(trainDetails)
    printStopNames(stops)
}

// -----------------------

const doc = `
${PROGRAM}
---------
Usage:
  ${PROGRAM} <from> <to> <dest_name>
  ${PROGRAM} -h | --help
  ${PROGRAM} --version

Options:
  -h --help               Show this screen.
  -V --version            Show version.

Examples
1. trains from RDG to PAD:
${PROGRAM} RDG PAD
`

const {docopt} = require('docopt');

// Can't use arguments here in strict mode or you get:
// "SyntaxError: Unexpected eval or arguments in strict mode"
// The names arguments and eval are special and forbidden as identifiers.
let args = docopt(doc, {
  version: `${VERSION} ${DATE} ${AUTHOR}`
})

const station = args['<from>']
const dest = args['<to>']
const dest_name = args['<dest_name>']
if (station && station.length === 3 && dest && dest.length === 3){
    getTrainsCallingAt(station,dest,dest_name)    // OK
}

File: trains.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# -*- coding: future_fstrings -*-
#
# trains.py
# (c) 2019 Mal Minhas, <mal@malm.co.uk>
#
# Licence
# --------
# Copyright 2019 Mal Minhas. All Rights Reserved.
#
# Description
# -----------
# transportAPI is documented here: https://developer.transportapi.com/docs?raml=https://transportapi.com/v3/raml/transportapi.raml
# More on three letter train codes here: http://www.railwaycodes.org.uk/crs/CRS0.shtm
# Some examples are:
# 1. London Paddington = PAD
# 2. London Waterloo = WAT
# 3. London Bridge = LD
#
# Installation
# ------------
# $ pip install -r requirements.txt
#
# Version
# -------
# 10.06.19  0.1   Initial version based on https://github.com/chrishutchinson/train-departure-screen/blob/master/src/trains.py
# 11.06.19  0.2   Added docopt support
# 

import requests

PROGRAM = __file__
VERSION = '0.2'
DATE = '11.06.19'
AUTHOR= 'Mal Minhas'

def readCred(fname):
    with open(fname,'r') as f:
        return f.read()
    raise f'Issue finding cred from {f}'

APP_ID = readCred('.transportAppId')
APP_KEY = readCred('.transportAppKey')

def getTrainsCallingAt(station_code,dest_code,dest_name):
    # station_code is 3 letter string.  eg. 'TWY','PADD'
    # from_offset is one hour in past by default
    # to_offset is two hours into future by default
    # type can be arrival|departure|pass
    url = f'http://transportapi.com/v3/uk/train/station/{station_code}/live.json'
    params = {'app_id': APP_ID, 'app_key': APP_KEY, 'station_code':station_code, 'calling_at': dest_code, 'type': 'departure'}
    #if station_name:
    #    params['station_name'] = station_name
    r = requests.get(url, params)
    data = r.json()
    data['destination_code'] = dest_code
    data['destination_name'] = dest_name
    return data

def procStop(stop):
    s = {}
    s['station_code'] = stop.get('station_code')
    s['station_name'] = stop.get('station_name')
    s['expected_arrival_time'] = stop.get('expected_arrival_time')
    s['platform'] = stop.get('platform')
    return s

def getStops(timetable_url,station_code,dest_code):
    ''' We need to make a GET request on the timetable URL to retrieve array of stops.  
    We are then only interested in a subset of the details fields for each stop.
    We also want to add whether that stop is on the designated journey or not.
    '''
    arr = []
    r = requests.get(timetable_url)
    stops = r.json().get('stops')
    on_route = False
    for stop in stops:
        s = procStop(stop)
        if s.get('station_code') == station_code:
            s['on_route'] = True
            on_route = True
        elif s.get('station_code') == dest_code:
            s['on_route'] = True
            on_route = False
        else:
            s['on_route'] = on_route
        arr.append(s)
    return arr

def formatHeader(d):
    header = f"==== Trains from {d.get('station_name')} ({d.get('station_code')}) to {d.get('destination_name')} "
    header += f"({d.get('destination_code')}) {d.get('time_of_day')} {d.get('date')} ===="
    return header

def formatDeparture(train,stops,d):
    source = [stop for stop in stops if stop.get('station_code') == d.get('station_code')][0]
    dest = [stop for stop in stops if stop.get('station_code') == d.get('destination_code')][0]
    route = [stop.get('station_name') for stop in stops if stop.get('on_route')]
    departure = f"{d.get('station_code')} {train.get('expected_departure_time')} -> {d.get('destination_code')}"
    departure += f" {dest.get('expected_arrival_time')} => {train.get('status')}\n"
    departure += f"\tTrain {train.get('train_uid')} ({train.get('operator')}) from {train.get('origin_name')}"
    departure += f" arriving at {d.get('station_name')} on platform {source.get('platform')}"
    departure += f" going to {d.get('destination_name')} platform {dest.get('platform')}. {len(route)} stops:"
    return departure

def formatTrains(d):
    # Keys: 'date', 'time_of_day', 'request_time', 'station_name', 'station_code', 'departures'
    # where 'departures' is a dict with one key 'all' which is a list of dicts of train departures
    printHeader(formatHeader(d))
    trains = d.get('departures').get('all')
    for train in trains:
        stops = getStops(train.get('service_timetable').get('id'),d.get('station_code'),d.get('destination_code'))
        trainDetails = formatDeparture(train,stops,d)
        printTrainDetails(trainDetails,stops)

def printTrainDetails(trainDetails,stops):
    print(trainDetails)
    printStopNames(stops)

def printHeader(header):
    print('='*len(header))
    print(header)
    print('='*len(header))

def printStopNames(stops):
    stopNames = ','.join([stop.get('station_name') for stop in stops if stop.get('on_route')])
    print(f'\t{stopNames}')


if __name__ == '__main__':
    import docopt

    usage = """
    {}
    ---------
    Usage:
    {} <from> <to> <dest_name>
    {} -h | --help
    {} -V | --version

    Options:
    -h --help               Show this screen.
    -V --version            Show version.

    Examples
    1. trains from RDG to PAD:
    {} RDG PAD
    
    """.format(
            *tuple([PROGRAM] * 5)
    )

    arguments = docopt.docopt(usage)
    # print(arguments)
    verbose = False
    force = False
    if arguments.get("--verbose") or arguments.get("-v"):
        verbose = True
    if arguments.get("--version") or arguments.get("-V"):
        print(f'{PROGRAM} version {VERSION}.  Author: {AUTHOR}')
    elif arguments.get("--help") or arguments.get("-h"):
        print(usage)
    else:
        source = arguments.get('<from>')
        dest = arguments.get('<to>')
        dest_name = arguments.get('<dest_name>')
        assert(len(source) == 3 and len(dest) == 3)
        trains = getTrainsCallingAt(source,dest,dest_name)     # OK
        formatTrains(trains)